src/ann_solo/utils.py

```diff
@@ -3,9 +3,200 @@
+                                                group_column='group')
+    else:
+        psms = mokapot.dataset.LinearPsmDataset(assert_columns_dtypes(full_dataset,mode), target_column='label',
+                                                spectrum_columns='ssm_id',
+                                                peptide_column='peptide')
+
+
+    #Train models
+    results, models = mokapot.brew(psms, model=model, test_fdr = fdr, folds=10, max_workers=os.cpu_count())
+
+    ##Compute q-value of the re-scored SSMs
+	best_targets = None
+    if mode == 'open':
+        ##Get full list of re-scored PSMs
+        rescored_pd = None
+        for key, value in results.group_confidence_estimates.items():
+            targets = value.confidence_estimates['psms']
+            decoys = value.decoy_confidence_estimates['psms']
+            rescored_pd = pd.concat([rescored_pd, targets, decoys], ignore_index=True)
+        #Sort table by group and score
+        rescored_pd.sort_values(['group', 'mokapot score'], ascending=False, inplace=True, ignore_index=True)
+        #Get a unique sorted list of group IDs
+        sorted_groups_list_rescored = sorted(list(set(rescored_pd['group'])), reverse=True)
+        #Compute the q-values and assign them to each group
+        full_q_value_list_rescored = []
+        for grp in sorted_groups_list_rescored:
+            full_q_value_list_rescored.extend(compute_q_value(rescored_pd[rescored_pd['group'] == grp]['label']))
+        rescored_pd['q_value'] = full_q_value_list_rescored
+        #Get targets table meeting the FDR threshold
+        best_targets = rescored_pd[(rescored_pd['q_value'] <= fdr) & (rescored_pd['label'] == True)]
+        #targets = results.confidence_estimates['psms']
+        #best_targets = targets[targets['mokapot q-value'] <= 0.01]
+    else:
+        targets = results.confidence_estimates['psms']
+        decoys = results.decoy_confidence_estimates['psms']
+        rescored_pd = pd.concat([targets, decoys], ignore_index=True)
+        ##Compute q-value
+        # Sort rows by score
+        rescored_pd.sort_values(['mokapot score'], ascending=False, inplace=True, ignore_index=True)
+        # Compute the q-values and assign them to each group
+        rescored_pd['q_value'] = compute_q_value(rescored_pd['label'])
+        # Get targets table meeting the FDR threshold
+        best_targets = rescored_pd[(rescored_pd['q_value'] <= fdr) & (rescored_pd['label'] == True)]
+
+    print("Number of matches found: {}".format(best_targets.shape[0]))
+    #Retreive final list of PSMs
+    for key in best_targets['ssm_id']:
+        psms_dictionary[key].search_engine_score = best_targets.loc[best_targets['ssm_id']== key,'mokapot score'].values[0]
+        psms_dictionary[key].q = best_targets.loc[best_targets['ssm_id'] == key, 'q_value'].values[0]
+        yield psms_dictionary[key]
+
+def compute_q_value(rankedSSMS: list) -> list:
```

Comment:
Use `pyteomics.auxiliary.qvalues(...)` instead, they have implemented all of this for us already. Look at the initial code in `utils.filter_fdr`.